Skip node ipset updates if iptables masquerading is disabled

pchaigno opened the pull request:
First commit introduces a helper function, the second skips all updates/creation of the node ipset if iptables-based masquerading is not used.

Fixes: https://github.com/cilium/cilium/issues/17711.

pchaigno committed:
option: Introduce IptablesMasqueradingIPvXEnabled helpers

These new functions will be used in subsequent commits to check if
iptables-based masquerading is enabled.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
iptables, node: Use ipset only with iptables masquerading

In commit 49cb2206 ("iptables: Don't masquerade traffic to cluster
nodes"), we introduced an ipset to track the set of node IP addresses
and avoid masquerading traffic to those IP in native routing mode.

This is however only needed if we're using iptables-based masquerading
(vs. BPF-based). When that's not the case, there's no reason to track
node IP addresses in this way, or even to require the ipset binary as a
dependency.

Fixes: 49cb2206 ("iptables: Don't masquerade traffic to cluster nodes")
Reported-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
iptables: Remove old ipsets

Remove old ipsets before installing new iptables rules, similarly to how
we remove old iptables rules and chains. The ipsets will then only be
created if we're using iptables-based masquerading. This ensures we
don't keep the ipsets when disabling iptables-based masquerading (e.g.,
when switching to BPF-based masquerading).

Signed-off-by: Paul Chaignon <paul@cilium.io>

jibi commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -492,8 +492,8 @@ func (m *IptablesManager) SupportsOriginalSourceAddr() bool {
-// removeOldRules removes iptables rules installed by Cilium.
-func (m *IptablesManager) removeOldRules(quiet bool) {
+// removeOldRulesAndIpsets removes iptables rules and ipsets installed by Cilium.
+func (m *IptablesManager) removeOldRulesAndIpsets(prefix string, quiet bool) {
```
Since this function has a `prefix` parameter, maybe it can be renamed to `removeRulesAndIpsets`? (And perhaps we can use a common prefix for both iptables and ipsets and pass the same prefix to `m.removeCiliumRules()` rather than using the `oldCiliumPrefix` constant?)

pchaigno replied:
Done and, since it was starting to look a bit too much like refactoring, I split it into a refactoring commit and the commit with changes.

pchaigno committed:
iptables: Refactor around a single OLD_ constant

In several places in the iptables code, we use an OLD_ string as a
prefix for iptables rules. This commit refactors the code a little to
ensure we use the same constant everywhere.

Suggested-by: Gilberto Bertin <gilberto@isovalent.com>
Signed-off-by: Paul Chaignon <paul@cilium.io>

jibi commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1339,7 +1339,7 @@ func (m *IptablesManager) installRules(ifName string) error {
 		}
 
-		if option.Config.EnableIPv4Masquerade && !option.Config.EnableBPFMasquerade {
+		if option.Config.IptablesMasqueradingEnabled() {
```
Wouldn't this cause IPv4 masquerading rules to be installed even if just IPv6 is enabled (and viceversa below)

pchaigno replied:
I don't think so. This code is conditioned by `option.Config.EnableIPv4` on line 1337. (Watch out for the incorrect GitHub indentations when viewing this directly on the pull request.)

twpayne commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -230,7 +230,7 @@ func (m *IptablesManager) removeCiliumRules(table string, prog iptablesInterface
 		// -A CILIUM_FORWARD -o cilium_host -m comment --comment "cilium: any->cluster on cilium_host forward accept" -j ACCEPT
 		// -A POSTROUTING -m comment --comment "cilium-feeder: CILIUM_POST" -j CILIUM_POST
-		if strings.Contains(rule, match) {
+		if strings.Contains(rule, prefix) {
```
Should this be `strings.HasPrefix` instead of `strings.Contains`?

twpayne commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1231,7 +1252,36 @@ func createIpset(name string, ipv6 bool) error {
+
+func renameIpset(name, prefix string) {
+	if !ipsetExists(name) {
+		return
```
Should this log a warning that the expected ipset does not exist?

Also, consider changing the name of the function to be `prefixIpsetName` or similar or add the prefix in the call to the function, rather than in the function. I would expect a function called `rename` to accept `oldname` and `newname` as arguments instead of `name` and `prefix`.

twpayne commented on pkg/option/config.go:
```diff
@@ -2157,6 +2157,12 @@ func (c *DaemonConfig) TunnelingEnabled() bool {
+// IptablesMasqueradingEnabled returns true if iptables-based
+// masquerading is enabled.
+func (c *DaemonConfig) IptablesMasqueradingEnabled() bool {
+	return !c.EnableBPFMasquerade && (c.EnableIPv4Masquerade || c.EnableIPv6Masquerade)
```
nit: as `||` has higher precedence than `&&`, the parentheses are not needed.

pchaigno replied:
Are you sure? `fmt.Println(false && false || true)` returns `true`.

qmonnet commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -497,19 +497,19 @@ func (m *IptablesManager) removeOldRules(quiet bool) {
 	tables := []string{"nat", "mangle", "raw", "filter"}
 	for _, t := range tables {
-		m.removeCiliumRules(t, ip4tables, oldCiliumPrefix)
+		m.removeCiliumRules(t, ip4tables, oldCiliumPrefix+"CILIUM_")
```
Should `"CILIUM_"` be a separate constant?

pchaigno replied:
Yeah, I considered it, but it's already hardcoded in so many places... I'm not sure I'm brave enough to make that change :sweat_smile: Plus, I'm not convinced there would be a lot of benefits.

qmonnet replied:
OK let's keep it as it is :+1:

qmonnet reviewed (approved):
All good for me!
